### PR TITLE
[ci] [gitlab] Add some tags for fast runner selection.

### PR DESCRIPTION
**Pull request by ejgallego**
This allows jobs to be picked by workers tagged `fast-runner`.

### CODE REVIEW

**Comment by ejgallego** [comment]
Results: https://gitlab.com/ejgallego/coq/pipelines/18681879/builds

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -52,7 +52,9 @@ before_script:
   - eval $(opam config env)
   - opam config list
   - opam install -j ${NJOBS} -y camlp5.${CAMLP5_VER} ocamlfind num ${EXTRA_OPAM}
-  - rm -rf ~/.opam/log/
+  - rm -rf ~/.opam/log/*
+  - rm -rf ~/.opam/archives/*
+  - rm -rf ~/.opam/${COMPILER}/build/*
```
Comment (SkySkimmer) [inline .gitlab-ci.yml]: What's this about?

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -261,14 +263,12 @@ ci-elpi:
...
-  allow_failure: true
-
-# ci-fiat-crypto:
-#   <<: *ci-template
-#   # out of memory error
-#   allow_failure: true
+  tags:
+    - fast-runner
+  # out of memory error
+  # allow_failure: true
```
Comment (SkySkimmer) [inline .gitlab-ci.yml]: You don't need to keep allow_failure comments.
Do comment the fast-runner tag on fiat-crypto with `runs out of memory on shared runners` or some such.

**Review by SkySkimmer — APPROVED** [review]
It's nice to have fiat-crypto in gitlab :)

**Comment by SkySkimmer** [comment]
What happens if you try to run it without having a fast-runner enabled?

**Comment by SkySkimmer** [comment]
>What happens if you try to run it without having a fast-runner enabled?

https://gitlab.com/SkySkimmer/coq/pipelines/18687569/builds
A warning flag appears next to the job with a tooltip on mousehover saying "Job is stuck. Check runners.".

**Comment by SkySkimmer** [comment]
Then at some point those jobs are marked failing.

File: .gitlab-ci.yml (hunk 1 above)
Comment (ejgallego) [reply .gitlab-ci.yml]: I was trying to improve caching time, now I see that .opam contains all 4 compilers.

Comment (ejgallego) [reply .gitlab-ci.yml]: Anyways this cleaning is convenient to do, but I'll move it to a separate commit.

**Comment by ejgallego** [comment]
> What happens if you try to run it without having a fast-runner enabled?

Yeah I think a tag makes mandatory for a job to execute in that runner. So this needs a bit more of thinking, but so far the experiment seems to have worked. Gitlab provides some tags such as 2GiB etc...

**Comment by ejgallego** [comment]
@SkySkimmer I propose this gets merged now.

**Comment by maximedenes** [comment]
@ejgallego I thought you wanted to move the clean up to a separate commit, as you wrote.

**Comment by maximedenes** [comment]
(I'm ok to merge either way, just let me know)

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
I think we should write somewhere (dev/ci/readme?) that gitlab CI users need a runner tagged `fast-runner` and how they can get one.

Does everyone who uses gitlab have the runner?

**Comment by SkySkimmer** [comment]
With @ejgallego's runners having issues I'm not sure about merging this PR (because it makes @ejgallego's runners required).
In any case it's also waiting on a doc update.

**Comment by ejgallego** [comment]
I agree, we need progress on the server side before merging this.

My "large" server is quite unstable unfortunately and while it has many cores each core is not that fast.

But the tags kinda worked as they should, so the experiment was OK. I've noticed that in fact Gitlab provides some tags such as `2gb`, could that mean those are faster workers?

**Comment by ejgallego** [comment]
It is clear that for Coq we need a build server which is good on CPU >> memory > cores.

**Comment by SkySkimmer** [comment]
>But the tags kinda worked as they should, so the experiment was OK. I've noticed that in fact Gitlab provides some tags such as 2gb, could that mean those are faster workers?

I don't know.

>It is clear that for Coq we need a build server which is good on CPU >> memory > cores.

We need around 4GB of memory / job for fiat-crypto, and having more is probably a bad thing since this way we detect if memory blows up very much (although it's no benchmark).

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
As mentioned in https://github.com/coq/coq/pull/7344#pullrequestreview-115134597 I don't want to require people have access to tagged runners to use gitlab CI, so please implement point 3 (gitlab-tagged.yml) from that comment.
(probably not in the main directory since people don't like having many files there)

**Comment by ejgallego** [comment]
A couple of time differences with fast-runner vs not:

| Name | Fast | GitLab Shared |
|------|------|---------------|
| `build` | 6m 7s | 14m 57s |
| `color` | 7m 13s  | 26m 34s |
| `formal-topology` | 17m 19s |  46m 08s |
| `geocoq` | 17m 09s | 28m 31s |
| `iris-lambdarust` | 25m 14s | 36m 33s |
| `math-comp` | 11m 52s | 27m 41s |
| `vst` | 0 | 37m 36s |

We can provide a default `fast-runner` for PRs easily, however on personal repositories this seems still a thorny issue, we may need to see if the tag can be made conditional.

**Comment by Zimmi48** [comment]
What is the issue with this solution?

> I don't want to require people have access to tagged runners to use gitlab CI, so please implement point 3 (gitlab-tagged.yml) from that comment.
(probably not in the main directory since people don't like having many files there)

**Comment by ejgallego** [comment]
I think we should be able to make this work without any special configuration on the user side.

**Comment by ejgallego** [comment]
For example, the following could work and allow a quick deployement:
```
fast-job:
  only:
    variables:
      - $CI_PROJECT_URL == "https://gitlab.com/coq/coq"

```

**Comment by SkySkimmer** [comment]
User configuration is far superior to hardcoding the main repo like that (and leaving everyone else with nothing).
Also that system simply doesn't work for the `build` jobs since others depend on `build` not `fast-build`

**Comment by ejgallego** [comment]
> User configuration is far superior to hardcoding the main repo like that (and leaving everyone else with nothing).

User configuration is still possible, but we cannot predict what users will do or what kind of runners they will have, so let's ship a default that works for everyone.

**Comment by SkySkimmer** [comment]
That's what I've been saying from the start

Gaëtan Gilbert

**Comment by ejgallego** [comment]
Yeah but I'd prefer to have a single or at worst a project variable as in the Windows case to enable the feature, anyways, I propose we follow here the same solution as in #7344 , whatever it is.

**Comment by ejgallego** [comment]
Indeed that's what I did not understand, why on #7344 we are going to do 2 and here we are going to do 3. IMHO it would make sense to have them controlled by the same mechanism.

**Comment by SkySkimmer** [comment]
>Indeed that's what I did not understand, why on #7344 we are going to do 2 and here we are going to do 3. IMHO it would make sense to have them controlled by the same mechanism.

Let's have an implementation of 3 then we can see what it looks like in practice.

**Comment by ejgallego** [comment]
> Let's have an implementation of 3 then we can see what it looks like in practice.

Umm, so I would basically include `.gitlab-ci.yml` and the override the variables, still thinking how to best do this without duplication.

**Comment by ejgallego** [comment]
No point on keeping this open until our CI server infrastructure shapes up.

**Pull request closed without merging**